lib/src/bootloader.rs

```diff
@@ -26,3 +26,94 @@ pub(crate) fn install_via_bootupd(
         .verbose()
         .run()
 }
+
+#[context("Installing bootloader using zipl")]
+pub(crate) fn install_via_zipl(device: &Utf8Path, boot_uuid: &str) -> Result<()> {
+    // Identify the target boot partition from UUID
+    let fs = crate::mount::inspect_filesystem_by_uuid(boot_uuid)?;
+    let boot_dir = Utf8Path::new(&fs.target);
+    let maj_min = fs.maj_min;
+
+    // Ensure that the found partition is a part of the target device
+    crate::blockdev::list_dev(device)?
+        .children
+        .with_context(|| format!("no partition found on {device:?}"))?
+        .iter()
+        .find(|dev| dev.maj_min == Some(maj_min.clone()))
+        .with_context(|| format!("partition device {:?} is not on {:?}", maj_min, device))?;
+
+    // Get an offset of the target partition by reading /sys/dev/block/<major:minor>/start
+    let sysfs_start_path = Utf8Path::new("/sys/dev/block").join(&maj_min).join("start");
+    let boot_offset = match std::fs::read_to_string(sysfs_start_path.clone()) {
+        Ok(v) => v.trim().parse::<i64>().map_err(|e| anyhow!(e)),
+        Err(e) if e.kind() == std::io::ErrorKind::NotFound => Ok(0), // partition-less disk
+        Err(e) => Err(anyhow!(e)),
+    }
+    .with_context(|| format!("reading {sysfs_start_path:?}"))?;
+
+    // Find the default BLS conf under /boot/loader/entries
+    // TODO: utilize the BLS parser in ostree
+    let bls_dir = boot_dir.join("boot/loader/entries");
+    let entries = bls_dir
+        .read_dir_utf8()?
+        .try_fold(Vec::new(), |mut acc, e| -> Result<_> {
+            let e = e?;
+            let name = Utf8Path::new(e.file_name());
+            if let Some("conf") = name.extension() {
+                acc.push(e.path().to_owned());
+            }
+            Ok(acc)
+        })?;
+    let bls_name = entries
+        .iter()
+        .max_by(|a, b| {
+            // TODO: Remove this unsafe code when a Rust version of strverscmp is available
+            #[link(name = "c")]
+            extern "C" {
+                fn strverscmp(s1: *const libc::c_char, s2: *const libc::c_char) -> libc::c_int;
```

Comment:
I thought this was a bit unusual and now I realize why we need to do it this way, it's because only GNU libc has `strverscmp` and Rust https://docs.rs/libc/latest/libc doesn't always expose those GNU-only APIs.

Reply:
No longer use `strverscmp`